pl022 spi v2

vchong opened the pull request:
Signed-off-by: Victor Chong <victor.chong@linaro.org>

vchong commented:
\+ @daniel-thompson fyi only

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,532 @@
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE) {
+			rdat = read32(pd->base + SSPDR);
+			DMSG("rdat: 0x%x\n", rdat);
```
relax cpu here ? (with a tee_time_wait or equiv)

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,532 @@
+
+static void pl022_sanity_check(struct pl022_data *pd)
+{
+	DMSG("SSPB2BTRANS: Expected: 0x2. Read: 0x%x\n",
```
could check here that  fields expected by `pl022_configure()` are initialized: non null, assuming pl022_init() resets the pl022_data structure. i.e `base`,  and `cs_gpio_pin`

vchong replied:
Something like this?
```
assert(pd &&
	pd->chip->ops &&
	pd->base &&
	pd->cs_gpio_base &&
	pd->clk_hz &&
	(pd->speed_hz && (pd->speed_hz <= (pd->clk_hz)/2)) &&
	(pd->mode && pd->mode <= SPI_MODE3) &&
	(pd->data_size_bits && pd->data_size_bits <= SSP_DATASIZE_MAX);
```

etiennec-linaro replied:
`mode` and `data_size_bits`do not need to be asserted. their value are checked in `pl022_configure()`.

vchong replied:
Isn't it easier to check/assert them here as you suggested than to wait till we get to `pl022_configure()`? If so, I can remove the value checkings in `pl022_configure()`, but should be no harm leaving them there?

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,532 @@
+
+void pl022_init(struct pl022_data *pd)
+{
+	pd->chip.ops = &pl022_ops;
```
`memset(*pd, 0, sizeof(*pd));` would insure caller properly set expected fields before calling `pl022_configure()`.

vchong replied:
This is just to set the `chip.ops` field. Caller could have set other fields already before calling `pl022_init`. `memset(*pd, 0, sizeof(*pd));` would wipe all of them away.

etiennec-linaro replied:
ok but warn caller that pl022_data must be kzalloced or at least memset to 0 before being filled.

a good thing could be `pl022_init` to alloc & reset `pl022_data` structure and init the internal stuff (as `chip.ops`) and return the data point.  A `pl022_release()` would be needed to properly free the structure and its resources (if any) but this might be just dead code.

vchong replied:
The assumption is that the caller will declare something like `static struct pl022_data xxx` and also set other fields (e.g. `xxx.data_size_bits = 8`, etc, to be later passed to and used by `pl022_configure` and friends) so we don't have to be concerned about alloc and free here. Am I missing a different use case perhaps? Maybe also add an assert to make sure pd is not null?

etiennec-linaro replied:
ok. so i think you don't need the `pl022_init()` and can set `chip.ops` from `pl022_configure()`.

vchong replied:
Ok. :) Thanks!

etiennec-linaro commented:
thanks for the instantiation of the spi devices. 

as a general minor comment: most of the `uint32_t` are simply unsigned values. I think you should better declare them as `unsigned int`, not caring about being 32b or 64b:
  - `i` and `j` counter in loops
  - `num_txpkts` and `num_rxpkts` arguments of send/recv handlers.
  - `clk_hz` and `speed_hz` fields of pl022_data structure,  and related `freqX` local variables.
 
also the `pl022_configuration()` should fail somehow in case of bad configuration. panic or return with some non-null value ?

jenswi-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,532 @@
+
+			/* rx 1 packet */
+			if (read32(pd->base + SSPSR) & SSPSR_RNE)
+				rdat[j++] = read8(pd->base + SSPDR);
```
How can we know that the buffer is large enough?

vchong replied:
The transfer by itself is always one packet/word in and one out, but since SPI doesn't specify rules for number of packets transmitted and/or received (does it?), the code was written with the assumption that there's a possibility to receive more packets than sent, and decisions regarding buffer size left to the caller assumed to have that knowledge, but yes, this does lead to a possibility of overflow.

What about
1. Read only as many packets as transmitted, or
2. Add a parameter to specify the length of the buffer, and only read up to its size, or
3. Other better suggestions?

jenswi-linaro replied:
Normally one would expect `*num_rxpkts` hold the number of max packets to receive.

vchong replied:
Ok, will use `num_rxpkts` as inout.

jenswi-linaro commented on core/include/spi.h:
```diff
@@ -0,0 +1,60 @@
+
+struct spi_ops {
+	void (*txrx8)(struct spi_chip *chip, uint8_t *wdat,
+		uint8_t *rdat, uint32_t num_txpkts, uint32_t *num_rxpkts);
```
`num_*` should be `size_t` instead of `uint32_t`

vchong replied:
Changed. Thanks!

jenswi-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,532 @@
+	uint16_t mode, data_size;
+	uint8_t cpsdvr, scr, lbm;
+
+	pl022_sanity_check(pd);
```
This function should do something if sanity checks fails.

vchong replied:
As mentioned previously, is the `assert` suggested in the sanity check function [above](https://github.com/OP-TEE/optee_os/pull/905#discussion_r70019435) sufficient?

etiennec-linaro replied:
no, below there are some error cases that also must be handle.

vchong replied:
All the asserts in sanity check above should make sure that there are no other errors below. Do you see any that are not taken care of?

Maybe only for `data_size_bits`?, in which case I can add asserts that it's 8 or 16 only.

etiennec-linaro replied:
yes, you can check them twice but once is sufficient and more easy to maintain i think.

vchong replied:
Ok, so are you ok then with just using asserts in sanity check and removing the error checkings here? Less code and easier to maintain as you said.

etiennec-linaro replied:
ok to assert all inputs.
But at least, config error cases for `mode` and `data_size_bits` switch/case blocks should call `panic()` instead of `EMSG()`.

vchong replied:
Ok, will change to `panic()`.

vchong commented:
> thanks for the instantiation of the spi devices.

np, thanks for the review! :)

>    i and j counter in loops
>    num_txpkts and num_rxpkts arguments of send/recv handlers.

Changed these to `size_t` per Jens.

>    clk_hz and speed_hz fields of pl022_data structure, and related freqX local variables.

Changed these to `unsigned int` as suggested.

> also the pl022_configuration() should fail somehow in case of bad configuration. panic or return with some non-null value ?

Is the `assert` in sanity check [above](https://github.com/OP-TEE/optee_os/pull/905#discussion_r70019435) sufficient?

jenswi-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,532 @@
+				tee_time_wait(1);
+
+			DMSG("TX FIFO available.. resuming..\n");
+			i--; /* retry current packet in next loop */
```
This is a bit complicated. Why not just test this stuff at the beginning of the loop instead of decreasing `i` to make another round.

vchong replied:
Ok. Will try to simplify this.

vchong commented:
Comments addressed. Please continue review.

Thanks!

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,505 @@
+	size_t i = 0, j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
```
in case no data to read/write, it could be better to not trig the chip-select low/high.
maybe it's not an issue with SPI.

vchong replied:
It's not an issue to toggle cs on no data as long as clk stays still. I can add a check to `return` if `num_txpkts == 0`, but that does seem to unnecessarily complicate the code for a use case that's probably close to 0%? I.e. why call read/write when there's nothing to read/write?

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,505 @@
+			&& (j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read16(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
```
I liked to previous that read/write at same time. just a matter of basic performances.

vchong replied:
That was my initial thought too, that read/write at the same time is better, but as @jenswi-linaro had pointed out, it made the code a bit more complicated than necessary. Performance-wise, it's also not necessarily better because previously we had write + read + is_fifo_full checking in one loop and then read in another loop, whereas now we just have a simple write and then read loop, with the fifo checking naturally integrated into the write loop. The new approach is also more conventional and inline with other pl022/spi drivers out there in the open.

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,505 @@
+	uint8_t *cpsdvr, uint8_t *scr)
+{
+	unsigned int freq1 = 0, freq2 = 0;
+	uint8_t tmp_cpsdvr1, tmp_scr1, tmp_cpsdvr2 = 0, tmp_scr2 = 0;
```
minor coding style matters:
- prefer 1 variable definition per line even if of same type.
- variables such as `tmp_cpsdvr1` only operates in the 1st `for` loop.
  `tmp_cpsdvr1` should be declared only inside this loop.

see here and in few other of the l022 functions.

vchong replied:
Ok. Will change. Thanks!

vchong replied:
Changes made to only declare 1 variable per line, but please note that the variables are used outside the `for` loop as well.

etiennec-linaro commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,505 @@
+	DMSG("set serial clock rate (scr), spi mode (phase and clock)\n");
+	DMSG("set frame format (spi) and data size (8- or 16-bit)\n");
+	io_mask16(pd->base + SSPCR0, SHIFT_U32(scr, 8) | mode | SSPCR0_FRF_SPI
+		| data_size, MASK_16);
```
minor coding style issue

the `|` (bitwise or) here should appear at the end of the upper line, rather than at start of the next line. This coding style directive stand for all operations: `+` `-` `&` `&&` and other operators.

for example:
```
if (condition_a_is_true &&
    condition_b_is_true)
	do_something();

value = (data & mask) | 
		new_value;
```

vchong replied:
Ok. Will change. Thanks!

etiennec-linaro commented:
`Reviewed-by: etienne carriere <etienne.carriere@linaro.org>`

vchong commented:
@jenswi-linaro Are you ok with the new changes as well?

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

vchong commented:
Rebased, squashed, tags added and ready for merge. Thanks!

jforissier commented:
I'd like to see the outcome of the Travis build. For some reason it has not triggered; I will investigate.
I found a compile error when `CFG_TEE_CORE_LOG_LEVEL=2`:
```
core/drivers/pl022_spi.c: In function ‘pl022_flush_fifo’:
core/drivers/pl022_spi.c:402:11: error: variable ‘rdat’ set but not used [-Werror=unused-but-set-variable]
  uint32_t rdat;
           ^
cc1: all warnings being treated as errors
mk/compile.mk:146: recipe for target 'out/arm-plat-hikey/core/drivers/pl022_spi.o' failed
```
(you need to add a `__maybe_unused` qualifier).

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+#include <trace.h>
+#include <util.h>
+
+/* spi register offsets */
```
SPI

vchong replied:
ok

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+#define SSPICR		0x020
+#define SSPDMACR	0x024
+
+/* hikey extensions */
```
HiKey

vchong replied:
ok

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+/* hikey extensions */
+#define SSPTXFIFOCR	0x028
+#define SSPRXFIFOCR	0x02C
+#define SSPB2BTRANS	0x030
```
What if the platform is not HiKey? Use of the above values should be conditional.

vchong replied:
Ok. As discussed offline, add `#define PLATFORM_hikey 1` in HiKey's `platform_config.h`, and add `#ifdef PLATFROM_hikey` here.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+#define SSPPCellID2	0xFF8
+#define SSPPCellID3	0xFFC
+
+/* spi register masks */
```
SPI

vchong replied:
ok

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+#define MASK_12 0xFFF
+#define MASK_8 0xFF
+#define MASK_4 0xF
+/* spi register masks */
```
SPI

vchong replied:
ok

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+
+static void pl022_print_peri_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 22 10 #4 0");
```
Is this correct? The below format will never print '#4'.

vchong replied:
`#` is supposed to denote an unknown number. Based on the pl022 spec, this is the only part of the peripheral ID that is not a fixed constant value. It returns the peripheral revision. Should I maybe change it to `22 10 [revision]4 0`?

jforissier replied:
Ah, I see. Personnally I would have used '?' instead, but that's not obviously better. Feel free to keep the '#'.

vchong replied:
I'll change to '?' anyway. '#' is maybe more common in the US but can be confusing. Thanks!

jforissier replied:
Yeah... this reminds me of an anecdote ;) when I was younger I was offered an Amiga computer by my parents. There was a vocal synthesis program coming with the OS. I remember it was reading "#" as "number" and that's how I learnt this character could be used for that purpose -- introduce, or stand for, a number.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+static void pl022_print_cell_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 0D F0 05 B1");
+	DMSG("Read: 0x %x %x %x %x\n",
```
Consistency: %X or "0x 0d f0 05 b1" above

d3zd3z replied:
Also, the expected line has leading zeros, perhaps %02x for each value would be more readable?

vchong replied:
Yes right. Is `Expected: 0x d f0 5 b1` ok, to keep it consistent with above?

jforissier replied:
I would find it more readable with %02x as @d3zd3z suggested (same above).

vchong replied:
%02x it is.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+	assert(pd->data_size_bits == 8 || pd->data_size_bits == 16);
+
+	DMSG("SSPB2BTRANS: Expected: 0x2. Read: 0x%x\n",
+		read32(pd->base + SSPB2BTRANS));
```
This is HiKey-specific as per the above comment so should be enabled for HiKey only

vchong replied:
As above, added `#ifdef PLATFROM_hikey`.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+		mode = PL022_SPI_MODE3;
+		break;
+	default:
+		EMSG("Invalid spi mode: %u\n", pd->mode);
```
SPI

vchong replied:
ok

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -0,0 +1,514 @@
+	}
+
+	DMSG("set serial clock rate (scr), spi mode (phase and clock)\n");
+	DMSG("set frame format (spi) and data size (8- or 16-bit)\n");
```
Acronyms should be uppercase (SCR, SPI)

vchong replied:
Yes sorry about that. I've changed them all to uppercase throughout, including some not mentioned here.

jforissier commented:
Please add `$make CFG_PL022=y` to `.travis.yml` so that this code gets compiled by the CI loop.

vchong commented:
`__maybe_unused` added to `rdat`

vchong commented:
Added `$make PLATFORM=hikey CFG_ARM64_core=y CFG_TEE_TA_LOG_LEVEL=4 DEBUG=1 CFG_PL061=y CFG_PL022=y` to `.travis.yml`.

vchong commented:
Comments addressed. Please continue review.

vchong commented:
Update: Added `CFG_PL061=y CFG_PL022=y` to some other platforms in `.travis.yml`.

jforissier commented on .travis.yml:
```diff
@@ -94,9 +94,11 @@ script:
 
   # FVP
   - $make PLATFORM=vexpress-fvp CFG_ARM32_core=y
+  - $make PLATFORM=vexpress-fvp CFG_ARM32_core=y CFG_PL061=y CFG_PL022=y
```
I'm wondering if we should enable PL061 and PL022 by default on HiKey instead. Then, no need to add anything to `.travis.yml` to have the code build-tested by Travis. Enabling them on other platforms would not check more things it seems. What do you think?

vchong replied:
Sounds good to me. Will change.

jforissier replied:
OK thanks, sorry for my bad suggestion in the first place ;)

vchong replied:
No problem. Updated and waiting for travis results.

jforissier commented on core/arch/arm/plat-hikey/conf.mk:
```diff
@@ -7,6 +7,8 @@ core_arm32-platform-aflags	+= -mfpu=neon
 $(call force,CFG_HWSUPP_MEM_PERM_PXN,y)
 $(call force,CFG_PL011,y)
+$(call force,CFG_PL061,y)
+$(call force,CFG_PL022,y)
```
That should be:
```
CFG_PL061 ?= y
CFG_PL022 ?= y
```
...at the end of the file. Because you don't want to force the values, i.e., 'make PLATFORM=hikey CFG_PL061=n CFG_PL022=n' should be allowed.

vchong replied:
Oops. Thanks! Fixed as well.

jforissier commented:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

vchong commented:
Rebased, squashed and additional tag added. Ok for merge? Thanks!
